text/0000-promise.md

```diff
@@ -0,0 +1,260 @@
+
+    std::thread::spawn(move || {
+        let result = perform_complex_operation();
+    
+        queue.send(move || {
+            let result = cx.number(result);
+
+            deferred.resolve(&cx, result);
+        });
+    });
+
+    Ok(promise)
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Before `Promise`, callbacks of the form `function (err, value)` were very common in JavaScript. Neon has excellent for for these "node style" callbacks in the `Task` trait.
+
+```rust
+fn fibonacci_async(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let n = cx.argument::<JsNumber>(0)?.value() as usize;
+    let cb = cx.argument::<JsFunction>(1)?;
+
+    FibonacciTask::new(n).schedule(cb);
+
+    Ok(cx.undefined())
+}
+```
+
+However, in idiomatic JavaScript, this should method should return a promise. This can be solved with some glue code in JavaScript:
+
+```js
+const util = require('util');
+const native = require('../native');
+
+export const fibonacciAsync = util.promisify(native.fibonacci_async);
+```
+
+Alternatively, a `Promise` can be constructed directly in Neon. Unlike an asynchronous method that accepts a callback, a method that returns a promise requires two parts:
+
+* A `Promise` to return from the method
+* A hook to resolve or reject the `Promise`
+
+In JavaScript, this looks like the following:
+
+```js
+function asyncMethod() {
+    return new Promise((resolve, reject) => {
+        if (Math.random() > 0.5) {
+            resolve();
+        } else {
+            reject();
+        }
+    });
+}
+```
+
+In `Neon`, when a `JsPromise` is constructed, a `Deferred` object is also provided. It can be thought of as the following pattern in JavaScript:
+
+```js
+function deferred() {
+    const deferred = {};
+    const promise = new Promise((resolve, reject) => {
+        deferred.resolve = resolve;
+        deferred.reject = reject;
+    });
+
+    return [promise, deferred];
+}
+
+function asyncMethod() {
+    const [promise, d] = deferred();
+
+    setTimeout(() => d.resolve(), 5000);
+
+    return promise;
+}
+```
+
+This could be written in Neon with the following:
+
+```rust
+fn async_method(cx: FunctionContext) -> JsResult<JsPromise> {
+    let queue = cx.queue();
+    let (promise, d) = cx.promise();
+
+    std::thread::spawn(move || {
+        std::thread::sleep(std::time::Duration::from_millis(5000));
+
+        queue.send(|cx| d.resolve(cx.undefined()));
+    });
+
+    Ok(promise)
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `JsPromise` API consists of two structs:
+
+* `JsPromise`. Opaque value; only useful for passing back to JavaScript.
+* `Deferred`. Handle for resolving and rejecting the related `Promise`.
+
+They may only be created with the `deferred` method on `Context`.
+
+```rust
+trait Context {
+    fn deferred(&mut self) -> (Handle<JsPromise>, Deferred);
```

Comment:
Elsewhere in the RFC you call this method `promise()`.

Reply:
Given the `let (sender, receiver) = channel();` precedent in std, should the order be `let (deferred, promise) = cx.promise();` instead?

Reply:
That's a great point on the order! I like that. 👍